compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -619,32 +622,59 @@ where
     recurse(tcx, ct, &mut f)
 }
 
+// Substitutes generics repeatedly to allow AbstractConsts to unify where a
+// ConstKind::Unevalated could be turned into an AbstractConst that would unify e.g.
+// Param(N) should unify with Param(T), substs: [Unevaluated("T2", [Unevaluated("T3", [Param(N)])])]
+#[inline]
+#[instrument(skip(tcx), level = "debug")]
+fn try_replace_substs_in_root<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mut abstr_const: AbstractConst<'tcx>,
+) -> Option<AbstractConst<'tcx>> {
+    while let Node::Leaf(ct) = abstr_const.root(tcx) {
+        match AbstractConst::from_const(tcx, ct) {
+            Ok(Some(act)) => abstr_const = act,
+            Ok(None) => break,
+            Err(_) => return None,
+        }
+    }
+
+    Some(abstr_const)
+}
+
 /// Tries to unify two abstract constants using structural equality.
+#[instrument(skip(tcx), level = "debug")]
 pub(super) fn try_unify<'tcx>(
     tcx: TyCtxt<'tcx>,
-    mut a: AbstractConst<'tcx>,
-    mut b: AbstractConst<'tcx>,
+    a: AbstractConst<'tcx>,
+    b: AbstractConst<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
 ) -> bool {
-    // We substitute generics repeatedly to allow AbstractConsts to unify where a
-    // ConstKind::Unevalated could be turned into an AbstractConst that would unify e.g.
-    // Param(N) should unify with Param(T), substs: [Unevaluated("T2", [Unevaluated("T3", [Param(N)])])]
-    while let Node::Leaf(a_ct) = a.root(tcx) {
-        match AbstractConst::from_const(tcx, a_ct) {
-            Ok(Some(a_act)) => a = a_act,
-            Ok(None) => break,
-            Err(_) => return true,
+    let a = match try_replace_substs_in_root(tcx, a) {
+        Some(a) => a,
+        None => {
+            return true;
         }
-    }
-    while let Node::Leaf(b_ct) = b.root(tcx) {
-        match AbstractConst::from_const(tcx, b_ct) {
-            Ok(Some(b_act)) => b = b_act,
-            Ok(None) => break,
-            Err(_) => return true,
+    };
+
+    let b = match try_replace_substs_in_root(tcx, b) {
+        Some(b) => b,
+        None => {
+            return true;
         }
-    }
+    };
```

Comment:
let else for both `try_replace_susts_in_root` here:

`let Some(a) = ... else { return true };